fish-rust/src/wutil/wrealpath.rs

```diff
@@ -0,0 +1,59 @@
+use std::fs::canonicalize;
+
+use crate::wchar::{wstr, WExt, WString};
+
+/// Wide character realpath. The last path component does not need to be valid. If an error occurs,
+/// `wrealpath()` returns `None`
+pub fn wrealpath(pathname: &wstr) -> Option<WString> {
+    if pathname.is_empty() {
+        return None;
+    }
+
+    let mut real_path = WString::new();
+    let mut narrow_path = pathname.to_owned();
+
+    // Strip trailing slashes. This is treats "/a//" as equivalent to "/a" if /a is a non-directory.
+    while narrow_path.len() > 1 && narrow_path.char_at(narrow_path.len() - 1) == '/' {
+        narrow_path.pop();
+    }
+
+    let mut narrow_result = canonicalize(narrow_path.to_string());
+
+    if let Ok(result) = narrow_result {
+        real_path.push_str(result.to_str()?);
+    } else {
+        // Check if everything up to the last path component is valid.
+        let mut pathsep_idx = narrow_path.find_last_char('/');
+
+        if let Some(0) = pathsep_idx {
+            // If the only pathsep is the first character then it's an absolute path with a
+            // single path component and thus doesn't need conversion.
+            real_path = narrow_path;
+        } else {
+            // Only call realpath() on the portion up to the last component.
+            if pathsep_idx.is_none() {
+                // If there is no "/", this is a file in $PWD, so give the realpath to that.
+                narrow_result = canonicalize(".");
+            } else {
+                // Only call realpath() on the portion up to the last component.
+                narrow_result = canonicalize(narrow_path[0..pathsep_idx.unwrap()].to_string());
+            }
+
+            if narrow_result.is_err() {
+                return None;
+            }
+
+            pathsep_idx = pathsep_idx.map(|idx| idx + 1);
```

Comment:
`None` corresponds to `npos` in the original code, which is defined as `-1` - thus `pathsep_idx++` is either `x + 1`, or `0` in case of `None`:

```suggestion
            let pathsep_idx = pathsep_idx.map_or(0, |idx| idx + 1);
```